presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java

```diff
@@ -190,11 +207,56 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     @Override
     public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session, ConnectorTableHandle table, Constraint<ColumnHandle> constraint, Optional<Set<ColumnHandle>> desiredColumns)
     {
-        InformationSchemaTableHandle handle = (InformationSchemaTableHandle) table;
-        ConnectorTableLayout layout = new ConnectorTableLayout(new InformationSchemaTableLayoutHandle(handle, constraint.getSummary()));
+        if (constraint.getSummary().isNone()) {
+            return ImmutableList.of();
+        }
+
+        InformationSchemaTableHandle handle = checkTableHandle(table);
+
+        Set<QualifiedTablePrefix> prefixes = calculatePrefixesWithSchemaName(session, constraint.getSummary(), constraint.predicate());
+        if (!handle.getSchemaTableName().equals(TABLE_SCHEMATA)) {
+            prefixes = calculatePrefixesWithTableName(session, prefixes, constraint.getSummary(), constraint.predicate());
+        }
+
+        ConnectorTableLayout layout = new ConnectorTableLayout(new InformationSchemaTableLayoutHandle(handle, prefixes));
         return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
     }
 
+    private Set<QualifiedTablePrefix> calculatePrefixesWithSchemaName(
+            ConnectorSession connectorSession,
+            TupleDomain<ColumnHandle> constraint,
+            Predicate<Map<ColumnHandle, NullableValue>> predicate)
+    {
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+        return metadata.listSchemaNames(session, catalogName).stream()
+                .filter(schema -> constraint.contains(SCHEMA_COLUMN_HANDLE, () -> utf8Slice(schema)))
```

Comment:
Can't we directly pass `utf8Slice(schema)` without supplier?

Reply:
Why do you ask `constraint` only about `table_schema`? What about the catalog?

Why do you ask `constraint` at all? Doesn't `predicate` represent the same filter?

> Can't we directly pass `utf8Slice(schema)` without supplier?

@sopel39  i guess the idea is to avoid slice construction when `constraint` is `ALL`. 

@kokosing instead of changing `TupleDomain` API, you could have a helper method here. We should have this in `TupleDomain` API when we have more usages.